python/aitemplate/backend/rocm/tensor/repeat.h

```diff
@@ -0,0 +1,188 @@
+ * such that the end result will have this head data
+ * repeated 1+num_repeat_copies
+ */
+__host__ hipError_t cuda_repeat_head(
+    void* data, ///< pointer to ROCM memory of size (at least)
+                ///< head_mem_bytes*(num_repeat_copies+1)
+    const size_t head_mem_bytes, ///< How many bytes to repeat
+    size_t num_repeat_copies, ///< How many times to repeat it (in addition to
+                              ///< the existing head data)
+    hipStream_t stream ///< ROCM Stream to use
+) {
+  hipError_t res = hipSuccess;
+  if (num_repeat_copies == 0)
+    return res;
+  if ((head_mem_bytes % 8) == 0) {
+    // no need to double memory any further if it is 64-bit aligned
+    res = cuda_repeat_head_vectorized(
+        static_cast<const int64_t* const>(data),
+        static_cast<int64_t*>(data) + (head_mem_bytes / 8),
+        head_mem_bytes / 8,
+        num_repeat_copies,
+        stream);
+    if (res != hipSuccess) {
+      return res;
+    }
+  } else {
+    res = hipMemcpyAsync(
+        static_cast<void*>(static_cast<uint8_t*>(data) + head_mem_bytes),
+        data,
+        head_mem_bytes,
+        hipMemcpyDeviceToDevice,
+        stream);
+    if (res != hipSuccess) {
+      return res;
+    }
+    if (num_repeat_copies >= 2) {
+      // recurse
+      // we have already repeated 1 time, therefore the (num_repeat_copies-1)
+      res = cuda_repeat_head(
+          data, head_mem_bytes * 2, (num_repeat_copies - 1) / 2, stream);
+      if (res != hipSuccess) {
+        return res;
+      }
+      // deal with possible remainder
+      if (((num_repeat_copies - 1) % 2) == 1) {
+        res = hipMemcpyAsync(
+            static_cast<void*>(
+                static_cast<uint8_t*>(data) +
+                num_repeat_copies * head_mem_bytes),
+            data,
+            head_mem_bytes,
+            hipMemcpyDeviceToDevice,
+            stream);
+      }
+    }
+  }
+  return res;
+}
+
+/**
+ * Repeatedly copy a source memory region into a target memory region
+ * such that the end result will have the source data
+ * repeated num_repeat_copies
+ */
+__host__ hipError_t cuda_repeat_src(
+    const void* const src, ///< Source memory region (readonly)
+    void* data, ///< Destination memory region (read/write, size of at least
+                ///< num_repeat_copies*head_mem_bytes)
+    const size_t head_mem_bytes, ///< Size of source region to copy
+    size_t num_repeat_copies, ///< How many times to copy the data from source
+                              ///< into data
+    hipStream_t stream ///< ROCM stream to use
+) {
+  hipError_t res = hipSuccess;
+  if (num_repeat_copies == 0) {
+    return res;
+  }
+
+  res = hipMemcpyAsync(
+      data, src, head_mem_bytes, hipMemcpyDeviceToDevice, stream);
+  if ((res != hipSuccess) || (num_repeat_copies == 1)) {
+    return res;
+  }
+  return cuda_repeat_head(data, head_mem_bytes, num_repeat_copies - 1, stream);
+}
```

Comment:
nit - could you add a newline at the end? Otherwise, our internal linter would raise an issue:

```
Files must end in a newline.
```

Thanks.